src/empty.js

```diff
@@ -1,21 +1,36 @@
 var _curry1 = require('./internal/_curry1');
-var _hasMethod = require('./internal/_hasMethod');
 
 
 /**
- * `empty` returns an empty list for any argument, except when the argument satisfies the
- * Fantasy-land Monoid spec. In that case, this function will return the result of invoking
- * `empty` on that Monoid.
+ * Returns the empty value of its argument's type. The argument must be a
+ * [Monoid](https://github.com/fantasyland/fantasy-land#monoid), Array,
+ * Object, or String.
  *
  * @func
  * @memberOf R
  * @category Function
- * @sig * -> []
- * @return {Array} An empty array.
+ * @sig Monoid m => m a -> m a
+ * @param {*} x
+ * @return {*}
  * @example
  *
- *      R.empty([1,2,3,4,5]); //=> []
+ *      R.empty(Just(42));      //=> Nothing()
+ *      R.empty([1, 2, 3]);     //=> []
+ *      R.empty('unicorns');    //=> ''
+ *      R.empty({x: 1, y: 2});  //=> {}
  */
 module.exports = _curry1(function empty(x) {
-  return _hasMethod('empty', x) ? x.empty() : [];
+  var repr = Object.prototype.toString.call(x);
+  switch (true) {
+    case x != null && typeof x.empty === 'function':
+      return x.empty();
+    case x != null && typeof x.constructor != null && typeof x.constructor.empty === 'function':
+      return x.constructor.empty();
+    case repr === '[object Array]':
+      return [];
+    case repr === '[object Object]':
+      return {};
+    case repr === '[object String]':
+      return '';
+  }
```

Comment:
I would personally prefer a nested ternary operator here. `switch (true)` is quite ugly in my opinion. Something along these lines:

```javascript
return x != null && typeof x.empty === 'function' ? x.empty()
     : repr === '[object Array]'  ? []
     : repr === '[object Object]' ? {}
                                  : undefined;
```